pilot/pkg/config/coredatamodel/controller.go

```diff
@@ -165,29 +169,57 @@ func (c *Controller) Apply(change *sink.Change) error {
 			return err
 		}
 
-		namedConfig, ok := innerStore[conf.Namespace]
-		if ok {
-			namedConfig[conf.Name] = conf
-		} else {
-			innerStore[conf.Namespace] = map[string]*model.Config{
-				conf.Name: conf,
+		switch {
+		case change.Incremental:
```

Comment:
IIUC what we need to do is to trigger a full update if the service portion of any of the services has changed.  So I think we need to loop through once and determine that.  Once we know if we're doing full or incremental, we can split up the logic into two branches.

Reply:
The full update is still taking place but not inside the loop (https://github.com/istio/istio/pull/12276/files#diff-babc5b7c836fa892988eef2d1544066bR221), this because EdsUpdate needs to happen on an individual config basis where as full update requires the entire configStore cache, so we do that once we receive all the configs.

Reply:
How about: for Service portion change, don't do incremental ? Since full config will be computed, sending the entire data set is not a huge problem ( we generate 100x more objects when distributing the config ).

Reply:
That's what the logic is exactly doing right now. It just gotten really confusing to have all the requirements in the same controller.

Reply:
> How about: for Service portion change, don't do incremental ? 

Wait, I think I misread what you meant, are you saying as soon as we encounter one service change in that collection, don't bother with `edsUpdate` and just do a full update? if so that seems totally reasonable.

Reply:
yeah exactly

Reply:
@nmittler I completely misread your comment, it makes total sense to do that.

Reply:
Yes, the only 'incremental' pilot supports at the moment is when an ServiceEntry has a change in endpoints and nothing else, i.e. result of K8S Service.Endpoints object update. 

On a stable cluster, that's the only change that will happen outside of config changes, and in a large cluster it's the main source of changes, that's why it's optimized with special code paths.